File: docker/worker-customcode-dotnet/harness/src/Honua.CustomCode.Harness/JobSpec.cs
// Copyright (c) Honua. All rights reserved.
// Licensed under the Elastic License 2.0. See LICENSE in the project root.

using System.Globalization;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace Honua.CustomCode.Harness;

/// <summary>Raised when the job inputs are missing or malformed.</summary>
public sealed class JobSpecException(string message) : Exception(message);

/// <summary>
/// Validated, fully-resolved job inputs (the harness half of the contract). Mirrors
/// the Python harness's <c>JobSpec</c> exactly — the same env vars and spec-file
/// precedence — so a single contract serves both runtimes.
/// </summary>
/// <remarks>
/// THE CONTRACT (image/harness half). The Batch container receives the job parameters
/// in one of two interchangeable ways, resolved in this order:
/// <list type="number">
/// <item>A mounted job-spec file — if <c>CUSTOMCODE_JOB_SPEC</c> points at a readable
/// JSON file, every field is read from it (preferred for large params/deps).</item>
/// <item>Discrete <c>CUSTOMCODE_*</c> environment variables otherwise.</item>
/// </list>
/// In BOTH modes the auth spine is always env-only and never in the spec file:
/// <c>HONUA_BASE_URL</c> and <c>HONUA_JOB_TOKEN</c> are read from the environment so the
/// scoped token follows the standard Batch secret-injection path and is strippable.
/// <para>
/// Field mapping (env var -&gt; spec key):
/// <c>CUSTOMCODE_RUNTIME</c> -&gt; runtime (e.g. "dotnet");
/// <c>CUSTOMCODE_REPO_URL</c> -&gt; repo_url;
/// <c>CUSTOMCODE_GIT_REF</c> -&gt; git_ref (40-hex SHA, validated);
/// <c>CUSTOMCODE_ENTRYPOINT</c> -&gt; entrypoint ("Assembly::Namespace.Type");
/// <c>CUSTOMCODE_DEPS_MANIFEST</c> -&gt; deps_manifest (the user .csproj, repo-relative);
/// <c>CUSTOMCODE_PARAMS_JSON</c> -&gt; params_json (inline JSON OR @path);
/// <c>CUSTOMCODE_OUTPUT_PREFIX</c> -&gt; output_prefix (s3://bucket/prefix);
/// <c>CUSTOMCODE_DECLARED_SCOPE</c> -&gt; declared_scope (advisory; comma/space list);
/// <c>CUSTOMCODE_OUTPUT_MAX_BYTES</c> -&gt; output_max_bytes (int; default 1 GiB).
/// </para>
/// </remarks>
public sealed class JobSpec
{
    /// <summary>The default output cap when the job declares none (1 GiB).</summary>
    public const long DefaultOutputMaxBytes = 1L * 1024 * 1024 * 1024;

    // A git_ref MUST be a fully-pinned 40-hex commit SHA. The server resolves to a
    // pinned SHA before dispatch; the harness re-validates as defense-in-depth.
    private static readonly Regex GitShaRegex = new("^[0-9a-f]{40}$", RegexOptions.CultureInvariant | RegexOptions.Compiled);

    // .NET entrypoint: Assembly::Namespace.Type. Matches the server-side validator.
    private static readonly Regex EntrypointRegex =
        new(@"^[A-Za-z_][A-Za-z0-9_]*::[A-Za-z_][A-Za-z0-9_]*(\.[A-Za-z_][A-Za-z0-9_]*)*$",
            RegexOptions.CultureInvariant | RegexOptions.Compiled);

    private JobSpec(
        string runtime,
        string repoUrl,
        string gitRef,
        string entrypoint,
        string depsManifest,
        JsonElement parameters,
        string outputPrefix,
        IReadOnlyList<string> declaredScope,
        long outputMaxBytes,
        string baseUrl,
        string jobToken)
    {
        Runtime = runtime;
        RepoUrl = repoUrl;
        GitRef = gitRef;
        Entrypoint = entrypoint;
        DepsManifest = depsManifest;
        Params = parameters;
        OutputPrefix = outputPrefix;
        DeclaredScope = declaredScope;
        OutputMaxBytes = outputMaxBytes;
        BaseUrl = baseUrl;
        JobToken = jobToken;
    }

    /// <summary>The runtime selector ("dotnet").</summary>
    public string Runtime { get; }

    /// <summary>The HTTPS git repository URL holding the user code.</summary>
    public string RepoUrl { get; }

    /// <summary>The pinned 40-hex commit SHA to check out.</summary>
    public string GitRef { get; }

    /// <summary>The entrypoint, "Assembly::Namespace.Type".</summary>
    public string Entrypoint { get; }

    /// <summary>The repo-relative path to the user's <c>.csproj</c>.</summary>
    public string DepsManifest { get; }

    /// <summary>The parsed <c>params_json</c> (tool-defined schema).</summary>
    public JsonElement Params { get; }

    /// <summary>The server-set per-job S3 output prefix.</summary>
    public string OutputPrefix { get; }

    /// <summary>The advisory declared scope (informational on the harness side).</summary>
    public IReadOnlyList<string> DeclaredScope { get; }

    /// <summary>The per-job output-size cap in bytes.</summary>
    public long OutputMaxBytes { get; }

    /// <summary>The Honua API base URL (from <c>HONUA_BASE_URL</c>).</summary>
    public string BaseUrl { get; }

    /// <summary>The scoped, job-bound bearer token (from <c>HONUA_JOB_TOKEN</c>).</summary>
    public string JobToken { get; }

    /// <summary>The assembly simple name from the entrypoint ("Asm" in "Asm::Type").</summary>
    public string EntrypointAssembly => Entrypoint.Split("::", 2)[0];

    /// <summary>The CLR type name from the entrypoint ("Type" in "Asm::Type").</summary>
    public string EntrypointType => Entrypoint.Split("::", 2)[1];

    /// <summary>Returns true iff <paramref name="value"/> is a fully-pinned 40-hex SHA.</summary>
    /// <param name="value">The candidate ref.</param>
    public static bool IsValidGitSha(string? value) => !string.IsNullOrEmpty(value) && GitShaRegex.IsMatch(value);

    /// <summary>Resolve and validate the job inputs from a spec file or env vars.</summary>
    /// <param name="env">The environment to read (defaults to the process environment).</param>
    /// <returns>The validated job spec.</returns>
    /// <exception cref="JobSpecException">When any required input is missing or malformed.</exception>
    public static JobSpec Load(IReadOnlyDictionary<string, string?>? env = null)
    {
        env ??= ReadProcessEnv();

        Func<string, string?> field;
        if (TryGet(env, "CUSTOMCODE_JOB_SPEC") is { } specPath)
        {
            var source = LoadSpecFile(specPath);
            field = key => source.TryGetValue(key, out var v) ? v : null;
        }
        else
        {
            // Spec keys map 1:1 to CUSTOMCODE_<UPPER> env vars.
            field = key => TryGet(env, "CUSTOMCODE_" + key.ToUpperInvariant());
        }

        var runtime = field("runtime") ?? "dotnet";
        var repoUrl = field("repo_url");
        var gitRef = field("git_ref");
        var entrypoint = field("entrypoint");
        var depsManifest = field("deps_manifest");
        var outputPrefix = field("output_prefix");

        // The auth spine is ALWAYS env-only — never sourced from the spec file.
        var baseUrl = TryGet(env, "HONUA_BASE_URL");
        var jobToken = TryGet(env, "HONUA_JOB_TOKEN");

        Require(repoUrl, "repo_url / CUSTOMCODE_REPO_URL");
        Require(gitRef, "git_ref / CUSTOMCODE_GIT_REF");
        Require(entrypoint, "entrypoint / CUSTOMCODE_ENTRYPOINT");
        Require(depsManifest, "deps_manifest / CUSTOMCODE_DEPS_MANIFEST");
        Require(outputPrefix, "output_prefix / CUSTOMCODE_OUTPUT_PREFIX");
        Require(baseUrl, "HONUA_BASE_URL");
        Require(jobToken, "HONUA_JOB_TOKEN");

        if (!IsValidGitSha(gitRef))
        {
            throw new JobSpecException(
                $"git_ref must be a 40-character hex commit SHA (got '{gitRef}'). " +
                "Branches, tags, and abbreviated SHAs are rejected.");
        }

        if (!EntrypointRegex.IsMatch(entrypoint!))
        {
            throw new JobSpecException(
                $"entrypoint must be 'Assembly::Namespace.Type' (got '{entrypoint}').");
        }

        if (!IsSafeRelativePath(depsManifest!))
        {
            throw new JobSpecException(
                $"deps_manifest must be a relative path within the repository (got '{depsManifest}').");
        }

        if (!StartsWithAny(repoUrl!, "https://", "git@", "ssh://", "http://"))
        {
            throw new JobSpecException($"repo_url has an unsupported scheme: '{repoUrl}'.");
        }

        if (!outputPrefix!.StartsWith("s3://", StringComparison.Ordinal))
        {
            throw new JobSpecException($"output_prefix must be an s3:// URI, got '{outputPrefix}'.");
        }

        return new JobSpec(
            runtime,
            repoUrl!,
            gitRef!,
            entrypoint!,
            depsManifest!,
            CoerceParams(field("params_json")),
            outputPrefix,
            CoerceScope(field("declared_scope")),
            CoerceMaxBytes(field("output_max_bytes")),
            baseUrl!,
            jobToken!);
    }

    private static IReadOnlyDictionary<string, string?> ReadProcessEnv()
    {
        var map = new Dictionary<string, string?>(StringComparer.Ordinal);
        foreach (System.Collections.DictionaryEntry kv in Environment.GetEnvironmentVariables())
        {
            map[(string)kv.Key] = kv.Value as string;
        }

        return map;
    }

    private static IReadOnlyDictionary<string, string?> LoadSpecFile(string path)
    {
        if (!File.Exists(path))
        {
            throw new JobSpecException($"CUSTOMCODE_JOB_SPEC points at a missing file: {path}");
        }

        JsonElement root;
        try
        {
            using var doc = JsonDocument.Parse(File.ReadAllText(path));
            root = doc.RootElement.Clone();
        }
        catch (Exception ex) when (ex is JsonException or IOException)
        {
            throw new JobSpecException($"Failed to read/parse job spec {path}: {ex.Message}");
        }

        if (root.ValueKind != JsonValueKind.Object)
        {
            throw new JobSpecException("Job spec file must contain a JSON object.");
        }

        var map = new Dictionary<string, string?>(StringComparer.Ordinal);
        foreach (var prop in root.EnumerateObject())
        {
            // params_json may be an embedded object; keep it as raw JSON text so the
            // common coercion path handles it like an inline JSON string.
            map[prop.Name] = prop.Value.ValueKind switch
            {
                JsonValueKind.String => prop.Value.GetString(),
                JsonValueKind.Null => null,
                _ => prop.Value.GetRawText(),
            };
        }

        return map;
    }

    private static JsonElement CoerceParams(string? raw)
    {
        if (string.IsNullOrEmpty(raw))
        {
            return ParseEmptyObject();
        }

        var text = raw;
        if (text.StartsWith('@'))
        {
            var paramsPath = text[1..];
            if (!File.Exists(paramsPath))
            {
                throw new JobSpecException($"params_json @file is missing: {paramsPath}");
            }

            text = File.ReadAllText(paramsPath);
        }

        try
        {
            using var doc = JsonDocument.Parse(text);
            if (doc.RootElement.ValueKind != JsonValueKind.Object)
            {
                throw new JobSpecException("params_json must decode to a JSON object.");
            }

            return doc.RootElement.Clone();
        }
        catch (JsonException ex)
        {
            throw new JobSpecException($"params_json is not valid JSON: {ex.Message}");
        }
    }

    private static JsonElement ParseEmptyObject()
    {
        using var doc = JsonDocument.Parse("{}");
        return doc.RootElement.Clone();
    }

    private static IReadOnlyList<string> CoerceScope(string? raw)
    {
        if (string.IsNullOrWhiteSpace(raw))
        {
            return [];
        }

        return raw
            .Split([',', ' ', '\t', '\n', '\r'], StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
            .ToArray();
    }

    private static long CoerceMaxBytes(string? raw)
    {
        if (string.IsNullOrEmpty(raw))
        {
            return DefaultOutputMaxBytes;
        }

        if (!long.TryParse(raw, NumberStyles.Integer, CultureInfo.InvariantCulture, out var value))
        {
            throw new JobSpecException($"output_max_bytes must be an integer, got '{raw}'.");
        }

        if (value <= 0)
        {
            throw new JobSpecException("output_max_bytes must be positive.");
        }

        return value;
    }

    private static bool IsSafeRelativePath(string path)
    {
        if (path.Contains('\0') || path.StartsWith('/') || path.StartsWith('\\'))
        {
            return false;
        }

        if (path.Length >= 2 && char.IsAsciiLetter(path[0]) && path[1] == ':')
        {
            return false;
        }

        foreach (var segment in path.Split('/', '\\'))
        {
            if (segment == "..")
            {
                return false;
            }
        }

        return true;
    }

    private static bool StartsWithAny(string value, params string[] prefixes)
        => prefixes.Any(p => value.StartsWith(p, StringComparison.Ordinal));

    private static string? TryGet(IReadOnlyDictionary<string, string?> env, string key)
        => env.TryGetValue(key, out var v) && !string.IsNullOrWhiteSpace(v) ? v.Trim() : null;

    private static void Require(string? value, string label)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            throw new JobSpecException($"Required job input is missing: {label}.");
        }
    }
}

File: docker/worker-customcode-python/harness/honua_customcode_harness/jobspec.py
"""Job-input parsing for the custom-code harness.

THE CONTRACT (image/harness half). The Batch container receives the job
parameters in **one of two interchangeable ways**, resolved in this order:

1. **A mounted job-spec file** — if ``CUSTOMCODE_JOB_SPEC`` points at a readable
   JSON file, every field is read from it. This is the preferred path for
   anything non-trivial (params_json/deps can be large; a file avoids env-size
   limits and keeps secrets out of ``/proc/<pid>/environ`` for the spec body).
2. **Discrete environment variables** — otherwise each field is read from a
   ``CUSTOMCODE_*`` env var. This is the cleanest path for the common small
   job and is what local/dev invocations use.

In BOTH modes the auth spine is always env-only and never in the spec file:
``HONUA_BASE_URL`` and ``HONUA_JOB_TOKEN`` are read from the environment so the
scoped token follows the standard Batch secret-injection path and is strippable
(see :mod:`honua_customcode_harness.harness`).

Field mapping (env var -> spec key):
    CUSTOMCODE_RUNTIME         -> runtime          (e.g. "python")
    CUSTOMCODE_REPO_URL        -> repo_url
    CUSTOMCODE_GIT_REF         -> git_ref          (40-hex SHA, validated)
    CUSTOMCODE_ENTRYPOINT      -> entrypoint       ("module:func")
    CUSTOMCODE_DEPS_MANIFEST   -> deps_manifest    (path, relative to repo root)
    CUSTOMCODE_PARAMS_JSON     -> params_json      (inline JSON string OR @path)
    CUSTOMCODE_OUTPUT_PREFIX   -> output_prefix    (s3://bucket/prefix)
    CUSTOMCODE_DECLARED_SCOPE  -> declared_scope   (advisory; comma/space list)
    CUSTOMCODE_OUTPUT_MAX_BYTES-> output_max_bytes (int; default 1 GiB)
"""

from __future__ import annotations

import json
import os
import re
from collections.abc import Mapping
from dataclasses import dataclass
from pathlib import Path
from typing import Any

# A git_ref MUST be a fully-pinned 40-hex commit SHA. Symbolic refs, tags,
# branches and abbreviated SHAs are rejected: the server resolves to a pinned
# SHA before dispatch and the harness re-validates as defense-in-depth.
_GIT_SHA_RE = re.compile(r"\A[0-9a-f]{40}\Z")

_DEFAULT_OUTPUT_MAX_BYTES = 1 * 1024 * 1024 * 1024  # 1 GiB

# The highest serving<->worker job-contract version this harness can run (ADR-0060
# principle #3b). Must stay in lockstep with the server's
# CustomCodeJobContract.ContractVersion (pinned by CustomCodeJobContractDriftTests).
# A submitted job whose injected HONUA_CONTRACT_VERSION exceeds this is rejected —
# the vX-server / vY-worker safety gate. Absent is treated as version 1 for back-compat.
SUPPORTED_CONTRACT_VERSION = 1
CONTRACT_VERSION_ENV = "HONUA_CONTRACT_VERSION"


class JobSpecError(ValueError):
    """Raised when the job inputs are missing or malformed."""


@dataclass(frozen=True)
class JobSpec:
    """Validated, fully-resolved job inputs (the harness half of the contract)."""

    runtime: str
    repo_url: str
    git_ref: str
    entrypoint: str
    deps_manifest: str | None
    params: Mapping[str, Any]
    output_prefix: str
    declared_scope: tuple[str, ...]
    output_max_bytes: int
    base_url: str
    job_token: str
    contract_version: int = 1

    @property
    def entrypoint_module(self) -> str:
        return self.entrypoint.split(":", 1)[0]

    @property
    def entrypoint_func(self) -> str:
        return self.entrypoint.split(":", 1)[1]


def is_valid_git_sha(value: str | None) -> bool:
    """Return True iff ``value`` is a fully-pinned 40-character lowercase SHA."""
    return bool(value) and bool(_GIT_SHA_RE.fullmatch(value or ""))


def validate_entrypoint(entrypoint: str | None) -> str:
    if not entrypoint or ":" not in entrypoint:
        raise JobSpecError(
            f"entrypoint must be 'module:func', got {entrypoint!r}."
        )
    module, _, func = entrypoint.partition(":")
    if not module or not func or "/" in module or " " in entrypoint:
        raise JobSpecError(f"entrypoint {entrypoint!r} is malformed.")
    return entrypoint


def _load_spec_file(path: str) -> Mapping[str, Any]:
    spec_path = Path(path)
    if not spec_path.is_file():
        raise JobSpecError(f"CUSTOMCODE_JOB_SPEC points at a missing file: {spec_path}")
    try:
        data = json.loads(spec_path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:  # pragma: no cover - io error
        raise JobSpecError(f"Failed to read/parse job spec {spec_path}: {exc}") from exc
    if not isinstance(data, Mapping):
        raise JobSpecError("Job spec file must contain a JSON object.")
    return data


def _coerce_params(raw: Any) -> Mapping[str, Any]:
    """Parse params_json into a dict.

    Accepts an already-parsed mapping (from a spec file), an inline JSON
    string, or an ``@/path/to/params.json`` reference.
    """
    if raw is None or raw == "":
        return {}
    if isinstance(raw, Mapping):
        return dict(raw)
    if isinstance(raw, str):
        text = raw
        if text.startswith("@"):
            params_path = Path(text[1:])
            if not params_path.is_file():
                raise JobSpecError(f"params_json @file is missing: {params_path}")
            text = params_path.read_text(encoding="utf-8")
        try:
            parsed = json.loads(text)
        except json.JSONDecodeError as exc:
            raise JobSpecError(f"params_json is not valid JSON: {exc}") from exc
        if not isinstance(parsed, Mapping):
            raise JobSpecError("params_json must decode to a JSON object.")
        return dict(parsed)
    raise JobSpecError(f"params_json has unsupported type {type(raw)!r}.")


def _coerce_scope(raw: Any) -> tuple[str, ...]:
    if not raw:
        return ()
    if isinstance(raw, (list, tuple)):
        items = [str(s).strip() for s in raw]
    else:
        items = re.split(r"[,\s]+", str(raw))
    return tuple(s for s in items if s)


def _coerce_contract_version(raw: Any) -> int:
    """Parse HONUA_CONTRACT_VERSION and fail closed if it exceeds this worker's max.

    Absent means a pre-versioning server (version 1). This is the defense-in-depth
    half of the submit-side contract gate (ADR-0060 #3b).
    """
    if raw is None or raw == "":
        return 1
    try:
        value = int(raw)
    except (TypeError, ValueError) as exc:
        raise JobSpecError(
            f"{CONTRACT_VERSION_ENV} must be a positive integer, got {raw!r}."
        ) from exc
    if value <= 0:
        raise JobSpecError(f"{CONTRACT_VERSION_ENV} must be a positive integer, got {raw!r}.")
    if value > SUPPORTED_CONTRACT_VERSION:
        raise JobSpecError(
            f"{CONTRACT_VERSION_ENV}={value} exceeds the version this worker can run "
            f"({SUPPORTED_CONTRACT_VERSION}). The worker image must be upgraded before it "
            "can run this job."
        )
    return value


def _coerce_max_bytes(raw: Any) -> int:
    if raw is None or raw == "":
        return _DEFAULT_OUTPUT_MAX_BYTES
    try:
        value = int(raw)
    except (TypeError, ValueError) as exc:
        raise JobSpecError(f"output_max_bytes must be an integer, got {raw!r}.") from exc
    if value <= 0:
        raise JobSpecError("output_max_bytes must be positive.")
    return value


def load_job_spec(env: Mapping[str, str] | None = None) -> JobSpec:
    """Resolve and validate the job inputs from a spec file or env vars.

    See module docstring for the precedence and field mapping.
    """
    env = os.environ if env is None else env

    if env.get("CUSTOMCODE_JOB_SPEC"):
        source: Mapping[str, Any] = _load_spec_file(env["CUSTOMCODE_JOB_SPEC"])

        def field(key: str) -> Any:
            return source.get(key)
    else:
        def field(key: str) -> Any:
            # Spec keys map 1:1 to CUSTOMCODE_<UPPER> env vars.
            return env.get(f"CUSTOMCODE_{key.upper()}")

    runtime = field("runtime") or "python"
    repo_url = field("repo_url")
    git_ref = field("git_ref")
    entrypoint = field("entrypoint")
    deps_manifest = field("deps_manifest") or None
    output_prefix = field("output_prefix")

    # Auth spine is ALWAYS env-only — never sourced from the spec file.
    base_url = env.get("HONUA_BASE_URL")
    job_token = env.get("HONUA_JOB_TOKEN")

    # Serving<->worker job-contract gate (ADR-0060 #3b), env-only like the auth spine.
    contract_version = _coerce_contract_version(env.get(CONTRACT_VERSION_ENV))

    _require(repo_url, "repo_url / CUSTOMCODE_REPO_URL")
    _require(git_ref, "git_ref / CUSTOMCODE_GIT_REF")
    _require(output_prefix, "output_prefix / CUSTOMCODE_OUTPUT_PREFIX")
    _require(base_url, "HONUA_BASE_URL")
    _require(job_token, "HONUA_JOB_TOKEN")

    if not is_valid_git_sha(git_ref):
        raise JobSpecError(
            f"git_ref must be a 40-character hex commit SHA (got {git_ref!r}). "
            "Branches, tags, and abbreviated SHAs are rejected."
        )
    validate_entrypoint(entrypoint)

    if not str(repo_url).startswith(("https://", "git@", "ssh://", "http://")):
        raise JobSpecError(f"repo_url has an unsupported scheme: {repo_url!r}.")
    if not str(output_prefix).startswith("s3://"):
        raise JobSpecError(f"output_prefix must be an s3:// URI, got {output_prefix!r}.")

    return JobSpec(
        runtime=str(runtime),
        repo_url=str(repo_url),
        git_ref=str(git_ref),
        entrypoint=str(entrypoint),
        deps_manifest=str(deps_manifest) if deps_manifest else None,
        params=_coerce_params(field("params_json")),
        output_prefix=str(output_prefix),
        declared_scope=_coerce_scope(field("declared_scope")),
        output_max_bytes=_coerce_max_bytes(field("output_max_bytes")),
        base_url=str(base_url),
        job_token=str(job_token),
        contract_version=contract_version,
    )


def _require(value: Any, label: str) -> None:
    if not value:
        raise JobSpecError(f"Required job input is missing: {label}.")
